sdk/python/kfp/local/dag_orchestrator.py

```diff
@@ -0,0 +1,316 @@
+        If DAG succeeds, a two-tuple of: (Status.SUCCESS, None).
+        If DAG fails, a two-tuple of: (Status.FAILURE, '<task-that-failed>').
+    """
+    from kfp.local import task_dispatcher
+
+    # prepare IOStore for DAG
+    dag_arguments_with_defaults = bind_defaults_to_dag_arguments(
+        dag_arguments=dag_arguments,
+        dag_inputs_spec=dag_component_spec.input_definitions,
+    )
+    for k, v in dag_arguments_with_defaults.items():
+        io_store.put_parent_input(k, v)
+
+    # execute tasks in order
+    dag_spec = dag_component_spec.dag
+    sorted_tasks = graph_utils.topological_sort_tasks(dag_spec.tasks)
+    while sorted_tasks:
+        task_name = sorted_tasks.pop()
+        component_name = dag_spec.tasks[task_name].component_ref.name
+        component_spec = components[component_name]
+        implementation = component_spec.WhichOneof('implementation')
+        # TODO: support pipeline-in-pipeline + control flow features
+        if implementation == 'dag':
+            raise ValueError(
+                'Control flow features and pipelines in pipelines are not yet supported by local pipeline execution.'
+            )
+        elif implementation != 'executor_label':
+            raise ValueError(
+                f'Got unknown component implementation: {implementation}')
+
+        executor_spec = executors[component_spec.executor_label]
+        validate_executor(executor_spec)
+        task_arguments = make_task_arguments(
+            task_inputs_spec=dag_spec.tasks[task_name].inputs,
+            io_store=io_store,
+        )
+
+        outputs, task_status = task_dispatcher._run_single_task_implementation(
+            pipeline_resource_name=pipeline_resource_name,
+            component_name=component_name,
+            component_spec=component_spec,
+            executor_spec=executor_spec,
+            arguments=task_arguments,
+            pipeline_root=pipeline_root,
+            runner=runner,
+            # let the outer pipeline raise the error
+            raise_on_error=False,
+            # components may consume input artifacts when passed from upstream
+            # outputs or parent component inputs
+            block_input_artifact=False,
+            # provide the same unique job id for each task for
+            # consistent placeholder resolution
+            unique_pipeline_id=unique_pipeline_id,
+        )
+
+        if task_status == status.Status.FAILURE:
+            return status.Status.FAILURE, task_name
+        elif task_status == status.Status.SUCCESS:
+            # update IO store when a task succeeds
+            for key, output in outputs.items():
+                io_store.put_task_output(
+                    task_name,
+                    key,
+                    output,
+                )
+        else:
+            raise ValueError(f'Got unknown task status: {task_status.name}')
+
+    return status.Status.SUCCESS, None
+
+
+def bind_defaults_to_dag_arguments(
```

Comment:
nit: this name is a bit misleading as function binds both user inputs and default values (when user inputs are not provided).